include/xgboost/tree_model.h

```diff
@@ -613,6 +626,107 @@ inline void RegTree::CalculateContributions(const RegTree::FVec& feat, unsigned
   out_contribs[split_index] += leaf_value - node_value;
 }
 
+// used by _compute_weight
+inline unsigned _seq_prod(unsigned a, unsigned b) {
+  unsigned out = 1.0;
+  for (unsigned i = a; i <= b; ++i) out *= i;
+  return out;
+}
+
+inline bst_float _compute_weight(unsigned num_fixed, unsigned num_integrated) {
```

Comment:
For depth 20 trees I got a factor of two gain by using a 20x20 lookup table, and swapping tuples for a struct. However, I also realized that the real problem is an unintentional exponential growth during recursion. This doesn't hurt shallow trees, but kills deep ones. Since the whole point of this algorithm was to avoid exponential running times I need to figure out how to fix this :) ...and I'll update the PR when I do.